doc/Authoring_Plugins.md

```diff
@@ -0,0 +1,188 @@
+
+#### Plug-ins
+
+pros:
+ + Code isolation: as plug-in's code in running in separate processes, it can't block Theia core processes.
+ + Can be loaded at runtime. No need to recompile the full IDE of Theia.
+ + Reduce compilation time
+ + Self-contained. A plug-in can be packaged into a single file and loaded directly after. No extra need to grab dependencies from npmjs, etc.
+ + Simple API
+   + no need to learn inversify or any framework.
+   + single entry point, with code completion to see possible calls with associated JsDoc.
+ + Upgrade easily from one theia version to another version as API is backward compliant.
+
+cons:
+ - Need to stick to this pre-defined API. It's not possible to tweak something if contribution point is not provided through API. Note that current API can be extended to support more stuff ;-)
+
+
+
+### Design
+A Theia app is composed of a core providing a set of widgets, commands, handlers, etc. for a specific functionality.
+
+Theia defines a runtime API allowing plug-ins to customize the IDE and add their behaviour to various aspects of the application. 
+
+In Theia, a plug-in has access to the API through an object named `theia` which is available in all plug-ins.
+[More details on API](https://github.com/theia-ide/theia/blob/master/packages/plugin/API.md).
+
+There are two natures of plug-ins:
+ - Backend plug-in. If you're familiar with VS Code extensions, it's very close. The plug-in's code is running in its own process on the server side. API is called and it's the API that will send some actions on user's browser/UI to register new commands, etc. All the callbacks are executed on the server side on a dedicated process.
+ - Frontend plug-in. In that case, callbacks are executed in a worker thread on the UI/browser. These plug-ins are only authorized to use "browser compliant" modules. For example opening or writing to a file is impossible as all the code of the plug-in is running on the browser side. But this approach is helpful if you really want to have some stuff on the client side to avoid some network operations.
+
+## Prerequisites
+
+Having a running Theia instance. (v0.3.12+)
+Instructions for getting Theia are available from [Theia repository](https://github.com/theia-ide/theia#getting-started).
+
+## Project Layout
+
+We are going to create a new project and for this we will create a folder named `theia-hello-world-plugin` that will contain the source code of the project.
+
+To ease the setup of such a repository, a [Yeoman code generator](https://www.npmjs.com/package/@theia/generator-plugin) exists and can be installed to scaffold the project.
+
+Generator can be installed and executed with the following commands. Note that these commands can be entered from a new terminal inside your running Theia instance.
+
+```bash
+npm install -g yo @theia/generator-plugin
+mkdir theia-hello-world-plugin
```

Comment:
One little clarification might help remove some potential confusion: at this point I was not sure where I should create the new plugin folder. Under the theia repo or on its own? Maybe mention abovethat it can be on it's own, independent of the theia repo?

Reply:
good point. I will clarify it.

Reply:
added note on that